src/librustc/traits/error_reporting.rs

```diff
@@ -640,16 +640,38 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                     ty::Predicate::ClosureKind(closure_def_id, kind) => {
                         let found_kind = self.closure_kind(closure_def_id).unwrap();
                         let closure_span = self.tcx.hir.span_if_local(closure_def_id).unwrap();
+                        let node_id = self.tcx.hir.as_local_node_id(closure_def_id).unwrap();
                         let mut err = struct_span_err!(
                             self.tcx.sess, closure_span, E0525,
                             "expected a closure that implements the `{}` trait, \
                                 but this closure only implements `{}`",
                             kind,
                             found_kind);
-                        err.span_note(
-                            obligation.cause.span,
-                            &format!("the requirement to implement \
-                                        `{}` derives from here", kind));
+
+                        let infer_tables = match self.tables {
+                            InferTables::Interned(tables) =>
+                                Some(InferTablesRef::Interned(tables)),
+                            InferTables::InProgress(tables) =>
+                                Some(InferTablesRef::InProgress(tables.borrow())),
+                            InferTables::Missing => None,
+                        };
+
+                        if let Some(tables) = infer_tables {
+                            if let Some(&(ty::ClosureKind::FnOnce, Some((span, name)))) =
```

Comment:
This message is specific to `FnOnce`, but similar scenarios can arise in other situations. For example, we could highlight the `x += 1` here and say something like "mutating this variable":

```rust
fn foo<F: Fn()>(f: F) { }
fn main() {
    let mut x = 0;
    foo(|| x += 1); // Error
}
```